src/github.com/getlantern/balancer/dialer.go

```diff
@@ -94,44 +78,94 @@ func (d *dialer) start() {
 					d.OnClose()
 				}
 				return
-			case t := <-d.errCh:
-				lastFailed = t
-			case <-timer.C:
-				ok := d.Check()
-				if ok {
-					lastCheckSucceeded = time.Now()
-					timer.Reset(longDuration)
-					consecCheckFailures = 0
-				} else {
-					consecCheckFailures += 1
-				}
+			case <-d.errCh:
+				d.markFailure()
+			case <-d.checkTimer.C:
+				go func() {
+					log.Tracef("Start checking dialer %s", d.Label)
+					ok := d.Check()
+					if ok {
+						d.markSuccess()
+					} else {
+						d.markFailure()
+					}
+				}()
 			}
 		}
 	}()
 }
 
-func (d *dialer) isActive() bool {
-	return atomic.LoadInt32(&d.active) == 1
+func (d *dialer) Stop() {
+	d.closeCh <- struct{}{}
+}
+
+func (d *dialer) AvgDialTime() int64 {
+	return atomic.LoadInt64(&d.avgDialTime)
+}
+func (d *dialer) ConsecSuccesses() int32 {
+	return atomic.LoadInt32(&d.consecSuccesses)
+}
+func (d *dialer) ConsecFailures() int32 {
+	return atomic.LoadInt32(&d.consecFailures)
+}
+
+func (d *dialer) CheckedDial(network, addr string) (net.Conn, error) {
```

Comment:
Oh, I see.  It might be less confusing to change CheckedDial -> Dial and Dial -> DoDial or UnderlyingDial or something like that.  Dial is the typical name used by things like net.Dialer, so someone looking at this is more likely to use Dial than CheckedDial since they probably don't know what CheckedDial means.